Avoid dispatch overhead in call to mkldnn convolution

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53327 [WIP] torch.utils.freeze
* #53908 [JIT] Add Reinplacing to MKLDNN Subgraphs
* #53326 use reshape when possible in broadcasting
* **#52614 Avoid dispatch overhead in call to mkldnn convolution**

This can speed up models by ~.5-1% from the base, which is much more  after they've been sped up 60% + with mkldnn.

Differential Revision: [D26696693](https://our.internmc.facebook.com/intern/diff/D26696693)

Commit:
Avoid dispatch overhead in call to mkldnn convolution

[ghstack-poisoned]

Commit:
Update on "Avoid dispatch overhead in call to mkldnn convolution"


This can speed up models by 5% (~.5-1% from the base, but ~5% after they've been sped up with mkldnn).






[ghstack-poisoned]

Commit:
Update on "Avoid dispatch overhead in call to mkldnn convolution"


This can speed up models by 5% (~.5-1% from the base, but ~5% after they've been sped up with mkldnn).






[ghstack-poisoned]

Commit:
Update on "Avoid dispatch overhead in call to mkldnn convolution"


This can speed up models by 5% (~.5-1% from the base, but ~5% after they've been sped up with mkldnn).






[ghstack-poisoned]

Commit:
Update on "Avoid dispatch overhead in call to mkldnn convolution"


This can speed up models by 5% (~.5-1% from the base, but ~5% after they've been sped up with mkldnn).






[ghstack-poisoned]

Commit:
Update on "Avoid dispatch overhead in call to mkldnn convolution"


This can speed up models by 5% (~.5-1% from the base, but ~5% after they've been sped up with mkldnn).






[ghstack-poisoned]

Commit:
Update on "Avoid dispatch overhead in call to mkldnn convolution"


This can speed up models by 5% (~.5-1% from the base, but ~5% after they've been sped up with mkldnn).






[ghstack-poisoned]

Krovatkin commented on torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:
```diff
@@ -77,6 +78,63 @@ Operation ConstantMKLDNNTensorOp(const Node* node) {
+          at::AutoNonVariableTypeMode non_var_type_mode(true);
+          // aten::convolution takes care of 0 dim case before calls into
+          // backends
+          if (input.size(0) == 0) {
```
it's weird that we only check `size(0)` but it seems you borrowed it from the convolution implementation.

eellison replied:
yep, this is coming from convolution.cpp

Krovatkin reviewed (approved):
:shipit:

Commit:
Update on "Avoid dispatch overhead in call to mkldnn convolution"


This can speed up models by 5% (~.5-1% from the base, but ~5% after they've been sped up with mkldnn).






[ghstack-poisoned]

Commit:
Update on "Avoid dispatch overhead in call to mkldnn convolution"


This can speed up models by 5% (~.5-1% from the base, but ~5% after they've been sped up with mkldnn).

Differential Revision: [D26696693](https://our.internmc.facebook.com/intern/diff/D26696693)

[ghstack-poisoned]

Commit:
Update on "Avoid dispatch overhead in call to mkldnn convolution"


This can speed up models by ~.5-1% from the base, which is much more  after they've been sped up 60% + with mkldnn.

Differential Revision: [D26696693](https://our.internmc.facebook.com/intern/diff/D26696693)

[ghstack-poisoned]

Commit:
Update on "Avoid dispatch overhead in call to mkldnn convolution"


This can speed up models by ~.5-1% from the base, which is much more  after they've been sped up 60% + with mkldnn.

Differential Revision: [D26696693](https://our.internmc.facebook.com/intern/diff/D26696693)

[ghstack-poisoned]

Commit:
Update on "Avoid dispatch overhead in call to mkldnn convolution"


This can speed up models by ~.5-1% from the base, which is much more  after they've been sped up 60% + with mkldnn.

Differential Revision: [D26696693](https://our.internmc.facebook.com/intern/diff/D26696693)

[ghstack-poisoned]

Commit:
Update on "Avoid dispatch overhead in call to mkldnn convolution"


This can speed up models by ~.5-1% from the base, which is much more  after they've been sped up 60% + with mkldnn.

Differential Revision: [D26696693](https://our.internmc.facebook.com/intern/diff/D26696693)

[ghstack-poisoned]

Commit:
Update on "Avoid dispatch overhead in call to mkldnn convolution"


This can speed up models by ~.5-1% from the base, which is much more  after they've been sped up 60% + with mkldnn.

Differential Revision: [D26696693](https://our.internmc.facebook.com/intern/diff/D26696693)

[ghstack-poisoned]

Commit:
Update on "Avoid dispatch overhead in call to mkldnn convolution"


This can speed up models by ~.5-1% from the base, which is much more  after they've been sped up 60% + with mkldnn.

Differential Revision: [D26696693](https://our.internmc.facebook.com/intern/diff/D26696693)

[ghstack-poisoned]

Commit:
Update on "Avoid dispatch overhead in call to mkldnn convolution"


This can speed up models by ~.5-1% from the base, which is much more  after they've been sped up 60% + with mkldnn.

Differential Revision: [D26696693](https://our.internmc.facebook.com/intern/diff/D26696693)

[ghstack-poisoned]
